File: problems/ts/15.matrix/dfs-traverse.ts
const directions = [
    [-1, 0],
    [0, +1],
    [+1, 0],
    [0, -1],
];

export function dfs<T>(
    matrix: T[][],
    row: number,
    col: number,
    seen: boolean[][],
    result: T[]
): void {
    if (row < 0 || col < 0 || row >= matrix.length || col >= matrix[0].length) {
        return;
    }
    result.push(matrix[row][col]);
    seen[row][col] = true;

    for (let i = 0; i < directions.length; ++i) {
        const direction = directions[i];
        if (
            matrix[row + direction[0]] &&
            matrix[row + direction[0]][col + direction[1]] &&
            !seen[row + direction[0]][col + direction[1]]
        ) {
            dfs(matrix, row + direction[0], col + direction[1], seen, result);
        }
    }
}

export function dfsTraverse<T>(matrix: T[][]) {
    const seenItems: boolean[][] = Array.from(
        { length: matrix.length },
        (_, index) => {
            return Array.from<boolean>({ length: matrix[index].length }).fill(
                false
            );
        }
    );
    const result: T[] = [];

    dfs(matrix, 0, 0, seenItems, result);

    return result;
}

File: golang/11.matrix/dfs/dfs_traverse.go
package matrix_dfs

var directions = [][]int{
	{-1, 0},
	{0, 1},
	{1, 0},
	{0, -1},
}

func dfs(matrix *[][]int, result *[]int, row int, col int, seen *[][]bool) {
	if row >= len(*matrix) || row < 0 || col >= len((*matrix)[0]) || col < 0 {
		return
	}

	if (*seen)[row][col] {
		return
	}

	*result = append(*result, (*matrix)[row][col])
	(*seen)[row][col] = true

	for _, direction := range directions {
		tempRow := row + direction[0]
		tempCol := col + direction[1]
		dfs(matrix, result, tempRow, tempCol, seen)
	}
}

func DfsTraverseMatrix(matrix [][]int) []int {
	seen := [][]bool{}
	result := []int{}

	for i := 0; i < len(matrix); i++ {
		seen = append(seen, []bool{})
		for j := 0; j < len(matrix[i]); j++ {
			seen[i] = append(seen[i], false)
		}
	}

	dfs(&matrix, &result, 0, 0, &seen)

	return result
}
